### ISSUE TITLE
Service of type ExternalName with port highjacks all traffic to given port

### ISSUE THREAD

**Issue by jcsorvasi** [istio/istio#20703]
**Bug description**

Creating a `Service` of type `ExternalName` with a port results in incorrect Envoy config, that will direct all traffic to said port regardless of host to this `Service`.

For example, creating the below `Service` will hijack all HTTPS traffic on the mesh towards this `Service`

```yaml
---
kind: Service
apiVersion: v1
metadata:
  name: my-externalname
  namespace: my-regular-ns # ns without Istio injection
spec:
  type: ExternalName
  externalName: bing.com
  ports:
  - port: 443
    targetPort: 443
```

As discussed in similar issues (#5580, #13349, #13479), creating this kind of service is generally not advisable, however it is perfectly accepted by Kubernetes.
Our team runs a 300+ node cluster hosting 1000+ Deployments with numerous teams using Kubernetes on a daily basis, ending up with this resource is not unlikely (as we've already seen in production).
We would like to gradually roll out Istio for teams to use, however currently we risk a single (non-mesh) `Service` breaking the entire service mesh.

Please advise.

**Affected product area**

**Expected behavior**

Creating such a `Service` does not hijack all traffic, thus breaking existing applications on the mesh.

**Steps to reproduce the bug**

Deploy test app onto the mesh

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test
  namespace: my-istio-enabled-ns # ns with Istio injection enabled
  labels:
    app: test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test
  template:
    metadata:
      labels:
        app: test
    spec:
      containers:
      - name: test
        image: nicolaka/netshoot
        args:
          - sleep
          - "30000000"
```

Exec into pod, initiate an HTTPS request to outside of the mesh (https://google.com)

```console
$ kubectl exec -it test-ff4df56f7-g6pzl -c test bash
bash-5.0# curl https://google.com
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="https://www.google.com/">here</A>.
</BODY></HTML>
bash-5.0#
```

Verify SSL certificate of https://google.com in the pod

```console
bash-5.0# echo | openssl s_client -showcerts -servername google.com -connect google.com:443 2>/dev/null | openssl x509 -inform pem -noout -text | grep "Subject Alternative Name" -A1
            X509v3 Subject Alternative Name:
                DNS:*.google.com, DNS:*.android.com, DNS:*.appengine.google.com, DNS:*.cloud.google.com, DNS:*.crowdsource.google.com, DNS:*.g.co, DNS:*.gcp.gvt2.com, DNS:*.gcpcdn.gvt1.com, DNS:*.ggpht.cn, DNS:*.gkecnapps.cn, DNS:*.google-analytics.com, DNS:*.google.ca, DNS:*.google.cl, DNS:*.google.co.in, DNS:*.google.co.jp, DNS:*.google.co.uk, DNS:*.google.com.ar, DNS:*.google.com.au, DNS:*.google.com.br, DNS:*.google.com.co, DNS:*.google.com.mx, DNS:*.google.com.tr, DNS:*.google.com.vn, DNS:*.google.de, DNS:*.google.es, DNS:*.google.fr, DNS:*.google.hu, DNS:*.google.it, DNS:*.google.nl, DNS:*.google.pl, DNS:*.google.pt, DNS:*.googleadapis.com, DNS:*.googleapis.cn, DNS:*.googlecnapps.cn, DNS:*.googlecommerce.com, DNS:*.googlevideo.com, DNS:*.gstatic.cn, DNS:*.gstatic.com, DNS:*.gstaticcnapps.cn, DNS:*.gvt1.com, DNS:*.gvt2.com, DNS:*.metric.gstatic.com, DNS:*.urchin.com, DNS:*.url.google.com, DNS:*.wear.gkecnapps.cn, DNS:*.youtube-nocookie.com, DNS:*.youtube.com, DNS:*.youtubeeducation.com, DNS:*.youtubekids.com, DNS:*.yt.be, DNS:*.ytimg.com, DNS:android.clients.google.com, DNS:android.com, DNS:developer.android.google.cn, DNS:developers.android.google.cn, DNS:g.co, DNS:ggpht.cn, DNS:gkecnapps.cn, DNS:goo.gl, DNS:google-analytics.com, DNS:google.com, DNS:googlecnapps.cn, DNS:googlecommerce.com, DNS:source.android.google.cn, DNS:urchin.com, DNS:www.goo.gl, DNS:youtu.be, DNS:youtube.com, DNS:youtubeeducation.com, DNS:youtubekids.com, DNS:yt.be
```

Dump listener config for `0.0.0.0:443`

```console
$ istioctl pc listeners test-77fd64bc5f-gdm7d --address 0.0.0.0 --port 443 -o json
[]
```

Everything checks out so far!

Create `Service` of type `ExternalName` with a port

```yaml
---
kind: Service
apiVersion: v1
metadata:
  name: my-externalname
  namespace: my-regular-ns
spec:
  type: ExternalName
  externalName: bing.com
  ports:
  - port: 443
    targetPort: 443
```

Exec into pod, initiate an HTTPS request to outside of the mesh (https://google.com)

```console
bash-5.0# curl https://google.com
curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number
```

Verify SSL certificate of https://google.com in the pod

```console
bash-5.0# echo | openssl s_client -showcerts -servername google.com -connect google.com:443 2>/dev/null | openssl x509 -inform pem -noout -text | grep "Subject Alternative Name" -A1
            X509v3 Subject Alternative Name:
                DNS:www.bing.com, DNS:dict.bing.com.cn, DNS:*.platform.bing.com, DNS:*.bing.com, DNS:bing.com, DNS:ieonline.microsoft.com, DNS:*.windowssearch.com, DNS:cn.ieonline.microsoft.com, DNS:*.origin.bing.com, DNS:*.mm.bing.net, DNS:*.api.bing.com, DNS:ecn.dev.virtualearth.net, DNS:*.cn.bing.net, DNS:*.cn.bing.com, DNS:ssl-api.bing.com, DNS:ssl-api.bing.net, DNS:*.api.bing.net, DNS:*.bingapis.com, DNS:bingsandbox.com, DNS:feedback.microsoft.com, DNS:insertmedia.bing.office.net, DNS:r.bat.bing.com, DNS:*.r.bat.bing.com, DNS:*.dict.bing.com.cn, DNS:*.dict.bing.com, DNS:*.ssl.bing.com, DNS:*.appex.bing.com, DNS:*.platform.cn.bing.com, DNS:wp.m.bing.com, DNS:*.m.bing.com, DNS:global.bing.com, DNS:windowssearch.com, DNS:search.msn.com, DNS:*.bingsandbox.com, DNS:*.api.tiles.ditu.live.com, DNS:*.ditu.live.com, DNS:*.t0.tiles.ditu.live.com, DNS:*.t1.tiles.ditu.live.com, DNS:*.t2.tiles.ditu.live.com, DNS:*.t3.tiles.ditu.live.com, DNS:*.tiles.ditu.live.com, DNS:3d.live.com, DNS:api.search.live.com, DNS:beta.search.live.com, DNS:cnweb.search.live.com, DNS:dev.live.com, DNS:ditu.live.com, DNS:farecast.live.com, DNS:image.live.com, DNS:images.live.com, DNS:local.live.com.au, DNS:localsearch.live.com, DNS:ls4d.search.live.com, DNS:mail.live.com, DNS:mapindia.live.com, DNS:local.live.com, DNS:maps.live.com, DNS:maps.live.com.au, DNS:mindia.live.com, DNS:news.live.com, DNS:origin.cnweb.search.live.com, DNS:preview.local.live.com, DNS:search.live.com, DNS:test.maps.live.com, DNS:video.live.com, DNS:videos.live.com, DNS:virtualearth.live.com, DNS:wap.live.com, DNS:webmaster.live.com, DNS:webmasters.live.com, DNS:www.local.live.com.au, DNS:www.maps.live.com.au
```

Dump listener config for 0.0.0.0:443

```console
$ istioctl pc listeners test-ff4df56f7-g6pzl --address 0.0.0.0 --port 443 -o json
[
    {
        "name": "0.0.0.0_443",
        "address": {
            "socketAddress": {
                "address": "0.0.0.0",
                "portValue": 443
            }
        },
        "filterChains": [
            {
                "filterChainMatch": {
                    "prefixRanges": [
                        {
                            "addressPrefix": "10.105.80.197",
                            "prefixLen": 32
                        }
                    ]
                },
                "filters": [
                    {
                        "name": "mixer",
                        "typedConfig": {
                            "@type": "type.googleapis.com/istio.mixer.v1.config.client.TcpClientConfig",
                            "transport": {
                                "networkFailPolicy": {
                                    "policy": "FAIL_CLOSE",
                                    "baseRetryWait": "0.080s",
                                    "maxRetryWait": "1s"
                                },
                                "checkCluster": "outbound|15004||istio-policy.istio-system.svc.cluster.local",
                                "reportCluster": "outbound|15004||istio-telemetry.istio-system.svc.cluster.local",
                                "reportBatchMaxEntries": 100,
                                "reportBatchMaxTime": "1s"
                            },
                            "mixerAttributes": {
                                "attributes": {
                                    "context.proxy_version": {
                                        "stringValue": "1.4.3"
                                    },
                                    "context.reporter.kind": {
                                        "stringValue": "outbound"
                                    },
                                    "context.reporter.uid": {
                                        "stringValue": "kubernetes://test-ff4df56f7-g6pzl.my-istio-enabled-ns"
                                    },
                                    "destination.service.host": {
                                        "stringValue": "my-externalname.my-regular-ns.svc.cluster.local"
                                    },
                                    "destination.service.name": {
                                        "stringValue": "my-externalname"
                                    },
                                    "destination.service.namespace": {
                                        "stringValue": "my-regular-ns"
                                    },
                                    "destination.service.uid": {
                                        "stringValue": "istio://my-regular-ns/services/my-externalname"
                                    },
                                    "source.namespace": {
                                        "stringValue": "my-istio-enabled-ns"
                                    },
                                    "source.uid": {
                                        "stringValue": "kubernetes://test-ff4df56f7-g6pzl.my-istio-enabled-ns"
                                    }
                                }
                            },
                            "disableCheckCalls": true
                        }
                    },
                    {
                        "name": "envoy.tcp_proxy",
                        "typedConfig": {
                            "@type": "type.googleapis.com/envoy.config.filter.network.tcp_proxy.v2.TcpProxy",
                            "statPrefix": "BlackHoleCluster",
                            "cluster": "BlackHoleCluster"
                        }
                    }
                ]
            },
            {
                "filters": [
                    {
                        "name": "mixer",
                        "typedConfig": {
                            "@type": "type.googleapis.com/istio.mixer.v1.config.client.TcpClientConfig",
                            "transport": {
                                "networkFailPolicy": {
                                    "policy": "FAIL_CLOSE",
                                    "baseRetryWait": "0.080s",
                                    "maxRetryWait": "1s"
                                },
                                "checkCluster": "outbound|15004||istio-policy.istio-system.svc.cluster.local",
                                "reportCluster": "outbound|15004||istio-telemetry.istio-system.svc.cluster.local",
                                "reportBatchMaxEntries": 100,
                                "reportBatchMaxTime": "1s"
                            },
                            "mixerAttributes": {
                                "attributes": {
                                    "context.proxy_version": {
                                        "stringValue": "1.4.3"
                                    },
                                    "context.reporter.kind": {
                                        "stringValue": "outbound"
                                    },
                                    "context.reporter.uid": {
                                        "stringValue": "kubernetes://test-ff4df56f7-g6pzl.my-istio-enabled-ns"
                                    },
                                    "destination.service.host": {
                                        "stringValue": "my-externalname.my-regular-ns.svc.cluster.local"
                                    },
                                    "destination.service.name": {
                                        "stringValue": "my-externalname"
                                    },
                                    "destination.service.namespace": {
                                        "stringValue": "my-regular-ns"
                                    },
                                    "destination.service.uid": {
                                        "stringValue": "istio://my-regular-ns/services/my-externalname"
                                    },
                                    "source.namespace": {
                                        "stringValue": "my-istio-enabled-ns"
                                    },
                                    "source.uid": {
                                        "stringValue": "kubernetes://test-ff4df56f7-g6pzl.my-istio-enabled-ns"
                                    }
                                }
                            },
                            "disableCheckCalls": true
                        }
                    },
                    {
                        "name": "envoy.tcp_proxy",
                        "typedConfig": {
                            "@type": "type.googleapis.com/envoy.config.filter.network.tcp_proxy.v2.TcpProxy",
                            "statPrefix": "outbound|443||my-externalname.my-regular-ns.svc.cluster.local",
                            "cluster": "outbound|443||my-externalname.my-regular-ns.svc.cluster.local",
                            "accessLog": [
                                {
                                    "name": "envoy.file_access_log",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/envoy.config.accesslog.v2.FileAccessLog",
                                        "path": "/dev/stdout",
                                        "format": "[%START_TIME%] \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%\" %RESPONSE_CODE% %RESPONSE_FLAGS% \"%DYNAMIC_METADATA(istio.mixer:status)%\" \"%UPSTREAM_TRANSPORT_FAILURE_REASON%\" %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% \"%REQ(X-FORWARDED-FOR)%\" \"%REQ(USER-AGENT)%\" \"%REQ(X-REQUEST-ID)%\" \"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\" %UPSTREAM_CLUSTER% %UPSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_REMOTE_ADDRESS% %REQUESTED_SERVER_NAME% %ROUTE_NAME%\n"
                                    }
                                }
                            ]
                        }
                    }
                ]
            },
            {
                "filterChainMatch": {
                    "applicationProtocols": [
                        "http/1.0",
                        "http/1.1",
                        "h2"
                    ]
                },
                "filters": [
                    {
                        "name": "envoy.http_connection_manager",
                        "typedConfig": {
                            "@type": "type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager",
                            "statPrefix": "outbound_0.0.0.0_443",
                            "rds": {
                                "configSource": {
                                    "ads": {}
                                },
                                "routeConfigName": "443"
                            },
                            "httpFilters": [
                                {
                                    "name": "mixer",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/istio.mixer.v1.config.client.HttpClientConfig",
                                        "transport": {
                                            "networkFailPolicy": {
                                                "policy": "FAIL_CLOSE",
                                                "baseRetryWait": "0.080s",
                                                "maxRetryWait": "1s"
                                            },
                                            "checkCluster": "outbound|15004||istio-policy.istio-system.svc.cluster.local",
                                            "reportCluster": "outbound|15004||istio-telemetry.istio-system.svc.cluster.local",
                                            "reportBatchMaxEntries": 100,
                                            "reportBatchMaxTime": "1s"
                                        },
                                        "serviceConfigs": {
                                            "default": {
                                                "disableCheckCalls": true
                                            }
                                        },
                                        "defaultDestinationService": "default",
                                        "mixerAttributes": {
                                            "attributes": {
                                                "context.proxy_version": {
                                                    "stringValue": "1.4.3"
                                                },
                                                "context.reporter.kind": {
                                                    "stringValue": "outbound"
                                                },
                                                "context.reporter.uid": {
                                                    "stringValue": "kubernetes://test-ff4df56f7-g6pzl.my-istio-enabled-ns"
                                                },
                                                "source.namespace": {
                                                    "stringValue": "my-istio-enabled-ns"
                                                },
                                                "source.uid": {
                                                    "stringValue": "kubernetes://test-ff4df56f7-g6pzl.my-istio-enabled-ns"
                                                }
                                            }
                                        },
                                        "forwardAttributes": {
                                            "attributes": {
                                                "source.uid": {
                                                    "stringValue": "kubernetes://test-ff4df56f7-g6pzl.my-istio-enabled-ns"
                                                }
                                            }
                                        }
                                    }
                                },
                                {
                                    "name": "envoy.cors"
                                },
                                {
                                    "name": "envoy.fault"
                                },
                                {
                                    "name": "envoy.router"
                                }
                            ],
                            "tracing": {
                                "operationName": "EGRESS",
                                "clientSampling": {
                                    "value": 100
                                },
                                "randomSampling": {
                                    "value": 100
                                },
                                "overallSampling": {
                                    "value": 100
                                }
                            },
                            "streamIdleTimeout": "0s",
                            "accessLog": [
                                {
                                    "name": "envoy.file_access_log",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/envoy.config.accesslog.v2.FileAccessLog",
                                        "path": "/dev/stdout",
                                        "format": "[%START_TIME%] \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%\" %RESPONSE_CODE% %RESPONSE_FLAGS% \"%DYNAMIC_METADATA(istio.mixer:status)%\" \"%UPSTREAM_TRANSPORT_FAILURE_REASON%\" %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% \"%REQ(X-FORWARDED-FOR)%\" \"%REQ(USER-AGENT)%\" \"%REQ(X-REQUEST-ID)%\" \"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\" %UPSTREAM_CLUSTER% %UPSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_REMOTE_ADDRESS% %REQUESTED_SERVER_NAME% %ROUTE_NAME%\n"
                                    }
                                }
                            ],
                            "useRemoteAddress": true,
                            "generateRequestId": true,
                            "upgradeConfigs": [
                                {
                                    "upgradeType": "websocket"
                                }
                            ],
                            "normalizePath": true
                        }
                    }
                ]
            }
        ],
        "deprecatedV1": {
            "bindToPort": false
        },
        "listenerFilters": [
            {
                "name": "envoy.listener.tls_inspector"
            },
            {
                "name": "envoy.listener.http_inspector"
            }
        ],
        "listenerFiltersTimeout": "0.100s",
        "continueOnListenerFiltersTimeout": true,
        "trafficDirection": "OUTBOUND"
    }
]
```

Notice how after creating the service, traffic to https://google.com got sent to bing.com

**Version**

Istio (v1.4.3)
```console
$ istioctl version --remote
client version: 1.4.3
control plane version: 1.4.3
data plane version: release-1.1-20191031-12-15 (7 proxies), 1.4.0 (10 proxies), 1.4.3 (8 proxies)
```

Kubernetes (v1.15.6)
```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.6", GitCommit:"7015f71e75f670eb9e7ebd4b5749639d42e20079", GitTreeState:"clean", BuildDate:"2019-11-13T11:20:18Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.6", GitCommit:"7015f71e75f670eb9e7ebd4b5749639d42e20079", GitTreeState:"clean", BuildDate:"2019-11-13T11:11:50Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```

Helm (v2.16.1)
```console
$ helm version
Client: &version.Version{SemVer:"v2.16.1", GitCommit:"bbdfe5e7803a12bbdf97e94cd847859890cf4050", GitTreeState:"clean"}
Error: could not find tiller
```

**How was Istio installed?**

Istio was installed following [Install with Helm via helm template](https://istio.io/docs/setup/install/helm/#option-1-install-with-helm-via-helm-template), by running `helm template (...) --values istio-globals.yaml --values istio.yaml` then `kubectl apply`

The contents of `istio-globals.yaml`

```yaml
global:
  arch:
    amd64: 3
    s390x: 0
    ppc64le: 0

  proxy:
    accessLogFile: "/dev/stdout"

  disablePolicyChecks: false

  sidecarInjectorWebhook:
    enabled: true
    rewriteAppHTTPProbe: false

  mtls:
    enabled: false
    auto: true

  controlPlaneSecurityEnabled: true

galley:
  enabled: true
  replicaCount: 2

security:
  enabled: true
  replicaCount: 2

sidecarInjectorWebhook:
  enabled: true
  replicaCount: 2

pilot:
  enabled: true
  autoscaleEnabled: true
  autoscaleMin: 2
  autoscaleMax: 5
  replicaCount: 2

  traceSampling: 100.0

  env:
    PILOT_SIDECAR_USE_REMOTE_ADDRESS: true

mixer:
  policy:
    enabled: true
    autoscaleEnabled: true
    autoscaleMin: 2
    autoscaleMax: 5
    replicaCount: 2

  telemetry:
    enabled: true
    autoscaleEnabled: true
    autoscaleMin: 2
    autoscaleMax: 5
    replicaCount: 2

  adapters:
    stdio:
      enabled: true

prometheus:
  enabled: false

grafana:
  enabled: false

tracing:
  enabled: false

kiali:
  enabled: false

gateways:
  enabled: false

```

The contents of `istio.yaml`

```yaml
global:
  defaultNodeSelector:
    mycompany.io/instance-class: system

  defaultTolerations:
    - key: "dedicated"
      operator: "Equal"
      value: "system"
      effect: "NoSchedule"

  defaultResources:
    requests:
      cpu: 1
      memory: 1Gi

kiali:
  enabled: true
  tag: v1.10.0
  contextPath: /
  dashboard:
    viewOnlyMode: true
  prometheusAddr: <REDACTED>
  grafanaURL: <REDACTED>
  ingress:
    enabled: true
    annotations:
      REDACTED: <REDACTED> # redacted external-dns.alpha.kubernetes.io and nginx.ingress.kubernetes.io annotations
    hosts:
      - <REDACTED>

```

**Environment where bug was observed (cloud vendor, OS, etc)**

Kubernetes deployed via kops on AWS

**Comment by howardjohn** [comment]
Agree, this seems pretty bad. If we can't support ExternalName we should just ignore them instead of doing what we do

**Comment by jcsorvasi** [comment]
@howardjohn Agreed! How can help? Any plans on how/when this will be addressed?

**Comment by howardjohn** [comment]
I am taking a look at some related externalname stuff today, I don't really know much about how its implemented. Ill update in a bit

**Comment by howardjohn** [comment]
very much related: https://github.com/istio/istio/issues/16458

**Comment by howardjohn** [comment]
I completely agree it's not good behavior the way it is, but you can also workaround this by naming the port http/https (I think) or removing the port entirely from the Service (it should still function the same). That does not solve the issue of some random person in your cluster breaking others (accidentally or intentionally)

**Comment by jcsorvasi** [comment]
We asked the tenant team to remove the port spec entirely from the `Service`, which resolved the issue, thank you. 
We're eagerly waiting for a solution to this though, as we cannot (in good conscience) encourage teams to move critical workloads to Istio while the mesh can be easily broken. 
Let us know if we can help with anything (testing early builds, etc.)!

**Comment by howardjohn** [comment]
@jcsorvasi I understand. For what its worth, you can scope this from "the whole mesh" to "some list of namespaces I chose" with https://istio.io/docs/reference/config/networking/sidecar/. 

I think the solution here is that we should not create a listener for any TCP ExternalName service. However, we can still create a cluster, I suppose, so they can do things like virtual service routing. However, I am not sure a great way to actually implement this that isn't super invasive

@ramaraochavali @hzxuzhonghu ideas?

**Comment by hzxuzhonghu** [comment]
Then the traffic to outside will through the passthrough cluster, it will miss some telemetries, right?

**Comment by howardjohn** [comment]
I guess so, yes, but it's better than blocking all traffic in the cluster?

**Comment by hzxuzhonghu** [comment]
Fair point

**Comment by hzxuzhonghu** [comment]
>However, we can still create a cluster, I suppose, so they can do things like virtual service routing.

To be honest, i can not understand this.

**Comment by zioalex** [comment]
Hi All, I do not see any solution here. @jcsorvasi how did you managed it in the end?

**Comment by jcsorvasi** [comment]
Hi @zioalex,

I'm not aware of a solution yet. We did not go to production with Istio yet, so this issue remains unsolved (and a potential issue) in our setup.

**Comment by BaudouinH** [comment]
Hello, my team and I encountered this issue on EKS 1.15 with Istio 1.6.11 (and Calico), while trying to connect to an out-of-cluster Redis database, on our staging environment.
We also had in another namespace a (no longer used) Service with an ExternalName for Redis, making it impossible for the client to connect to the db.

We found two possible work-around (instead of deleting the Service with an ExternalName) : 
-  Creating a ServiceEntry for the destination (here the Redis db), with or without specifying the protocol (both worked)
- Or, excluding the Outbound port from the mesh

**Comment by renaudhager** [comment]
Hi guys,

I have encountered an issue that might be related to this one.
In my case, it's the envoy metadata that are highjacked by a headless service.
I have a service named backend-a defined like this:
```
---
apiVersion: v1
kind: Service
metadata:
  name: backend-a
  namespace: test
  labels:
    app: backend-a
    service: backend-a
spec:
  type: ClusterIP
  ports:
  - port: 80
    targetPort: 5000
    name: http
  selector:
    app: backend-a
```
And another one defined like this:
```
apiVersion: v1
kind: Service
metadata:
  name: node-exporter
  namespace: lens-metrics
  annotations:
    prometheus.io/scrape: 'true'
spec:
  type: ClusterIP
  clusterIP: None
  selector:
    name: node-exporter
    phase: prod
  ports:
    - name: metrics
      protocol: TCP
      port: 80
      targetPort: 9100
```

From the POD in `test` namespace if we curl http://backend-a.test.svc.cluster.local envoy identify this request to `node-exporter`.

Configured service:
```
k get service -A | grep "80/TCP"
istio-system           tracing                           ClusterIP      172.20.139.202   <none>        80/TCP                                                       5d20h
kube-system            kube-state-metrics                ClusterIP      None             <none>        8080/TCP,8081/TCP                                            16d
kube-system            kubernetes-dashboard              ClusterIP      172.20.55.205    <none>        80/TCP                                                       48d
kube-system            traefik-ingress-private-service   ClusterIP      172.20.55.26     <none>        80/TCP,8080/TCP                                              48d
kube-system            traefik-ingress-public-service    ClusterIP      172.20.252.205   <none>        80/TCP,8080/TCP                                              48d
lens-metrics           kube-state-metrics                ClusterIP      172.20.120.30    <none>        8080/TCP                                                     15d
lens-metrics           node-exporter                     ClusterIP      None             <none>        80/TCP                                                       15d
lens-metrics           prometheus                        ClusterIP      172.20.225.232   <none>        80/TCP                                                       15d
test                   backend-a                         ClusterIP      172.20.234.100   <none>        80/TCP                                                       5d19h
```

PODs state and IP:
```
$ k get po -n test -o wide
NAME                            READY   STATUS    RESTARTS   AGE   IP           NODE                                       NOMINATED NODE   READINESS GATES
backend-a-v1-57b6f5f658-tx4pw   2/2     Running   0          8h    10.1.17.96   ip-10-1-17-53.eu-west-1.compute.internal   <none>           <none>
backend-a-v2-68974cb9f-42qzh    2/2     Running   0          10h   10.1.1.165   ip-10-1-1-51.eu-west-1.compute.internal    <none>           <none>
backend-b-5d75484db7-ncnpp      2/2     Running   0          10h   10.1.1.63    ip-10-1-1-51.eu-west-1.compute.internal    <none>           <none>
frontend-v1-5d6c4d89d8-6lhpg    2/2     Running   0          17m   10.1.3.253   ip-10-1-3-149.eu-west-1.compute.internal   <none>           <none>
frontend-v2-69d4b5c9f7-86sv2    2/2     Running   0          17m   10.1.15.67   ip-10-1-1-118.eu-west-1.compute.internal   <none>           <none>
```

Header from the `backend-a` app:
```
{
  "Accept": "*/*",
  "Content-Length": "0",
  "Host": "backend-a.test.svc.cluster.local",
  "User-Agent": "curl/7.35.0",
  "X-B3-Sampled": "1",
  "X-B3-Spanid": "7ec64f84b13ec01d",
  "X-B3-Traceid": "d3b7091b04ac2ea97ec64f84b13ec01d",
  "X-Forwarded-Proto": "http",
  "X-Request-Id": "0ac58168-23a3-94d2-8ea2-ba3c4e816e6e"
}
```

Envoy logs on the POD where curl is executed:
```
{
	"id": "AQAAAXaotQANj98XKAAAAABBWGFvdFE2dEFBQUxqbFJvZ3FEYXBnQUE",
	"content": {
		"timestamp": "2020-12-28T09:36:43.021Z",
		"tags": [
			<REDACTED>
		],
		"host": "eks-worker-i-0518ea51a9e3ddbbb",
		"service": "proxyv2",
		"attributes": {
			"response_code": "0",
			"method": "-",
			"route_name": "-",
			"downstream_remote_address": "10.1.3.253:46360",
			"requested_server_name": "-",
			"upstream_service_time": "-",
			"bytes_received": "103",
			"bytes_sent": "522",
			"x_forwarded_for": "-",
			"istio_policy_status": "-",
			"duration": "7",
			"response_flags": "-",
			"path": "-",
			"start_time": "2020-12-28T09:36:38.886Z",
			"protocol": "-",
			"upstream_cluster": "outbound|80||node-exporter.lens-metrics.svc.cluster.local",
			"authority": "-",
			"upstream_local_address": "10.1.3.253:46362",
			"downstream_local_address": "172.20.234.100:80",
			"upstream_host": "172.20.234.100:80",
			"upstream_transport_failure_reason": "-",
			"request_id": "-",
			"user_agent": "-"
		}
	}
}
```

Envoy logs from the backend-a app:
```
{
  "id": "AQAAAXaoqNWgotEUeQAAAABBWGFvcU9La0FBQnAxSE1tYXROU09nQUg",
  "content": {
    "timestamp": "2020-12-28T09:23:25.728Z",
    "tags": [
      <REDACTED>
    ],
    "host": "eks-worker-i-0998657dd903c37bc",
    "service": "proxyv2",
    "attributes": {
      "response_code": "200",
      "method": "GET",
      "route_name": "default",
      "downstream_remote_address": "10.1.3.253:43048",
      "requested_server_name": "-",
      "upstream_service_time": "2",
      "bytes_received": "0",
      "bytes_sent": "290",
      "x_forwarded_for": "-",
      "istio_policy_status": "-",
      "duration": "2",
      "response_flags": "-",
      "path": "/headers",
      "start_time": "2020-12-28T09:23:20.161Z",
      "protocol": "HTTP/1.1",
      "upstream_cluster": "inbound|80|http|backend-a.test.svc.cluster.local",
      "authority": "backend-a.test.svc.cluster.local",
      "upstream_local_address": "127.0.0.1:40312",
      "downstream_local_address": "10.1.17.96:5000",
      "upstream_host": "127.0.0.1:5000",
      "upstream_transport_failure_reason": "-",
      "request_id": "0ac58168-23a3-94d2-8ea2-ba3c4e816e6e",
      "user_agent": "curl/7.35.0"
    }
  }
}
```

We clearly observe that on the source Envoy proxy, the `upstream_cluster` associated to the request is `node-exporter.lens-metrics.svc.cluster.local` (which is wrong) but the `downstream_local_address` is correct (172.20.234.100).

The result of this is that in Kiala I have some traffic from `unknow` hitting the backend-a app.

![image](https://user-images.githubusercontent.com/12563305/103558862-3460cd00-4ead-11eb-8a86-ff9694e760fa.png)

If I remove or change the port of the service `node-exporter` everything works fine.

**Comment by howardjohn** [comment]
Not stale

**Comment by malcolmrebughini** [comment]
@howardjohn are there any updates on this issue?

**Comment by hzxuzhonghu** [comment]
With dns proxy, we could allocate an ip for the external service, and then we can handle this case. The tricky point is that dns proxy is optinal.

**Comment by shikanon** [comment]
the istio 1.8.5 also has this problem, when I create the `ExternalName` service, all traffic will be highjacks.
```bash
$ istioctl pc listeners tools-redis-df69c5598-jd2nr.rcmd-tt --address 0.0.0.0 --port 6379
ADDRESS PORT MATCH                        DESTINATION
0.0.0.0 6379 Trans: raw_buffer; App: HTTP Route: 6379
0.0.0.0 6379 ALL                          Cluster: outbound|6379||redis-zt-channel-filter.databases.svc.cluster.local
```
the istio version:
```bash
$ istioctl version --remote
client version: 1.8.5
control plane version: 1.8.4
data plane version: 1.8.4 (44 proxies)
```

**Comment by wufengzlw** [comment]
> With dns proxy, we could allocate an ip for the external service, and then we can handle this case. The tricky point is that dns proxy is optinal.

@hzxuzhonghu 
Yes, if our external service has a fixed IP, we can enable DNS proxy and create a serviceEntry for that.
Any idea for external service with DNS name instead of IP? such as a RDS in AWS

**Comment by howardjohn** [comment]
not stale

**Comment by hzxuzhonghu** [comment]
@wufengzlw Sorry, I am not familiar with AWS RDS

**Comment by rpuserh** [comment]
Facing the same problem. What is the solution, to not specify ports for external service?

**Comment by DingGGu** [comment]
We use a policy controller such as [Kyverno](https://kyverno.io/) to enforce users to specify a port in ExternalName to avoid breaking the entire network of the cluster.

If Istio doesn't need to handle ExternalName, I'd like to ignore it or port specifics.

**Comment by idouba** [comment]
not stale

**Comment by hzxuzhonghu** [comment]
/not stale

Lets see what can we do to prevent such critical result

**Comment by idouba** [comment]
> With dns proxy, we could allocate an ip for the external service, and then we can handle this case. The tricky point is that dns proxy is optinal.

Yes，Just like allocate a clusterIP for No vip ServiceEntry, is it possible to allocate a IP for specified externalName service by Istio DNS, and make Envoy listen on this Ip to prevent blocking other same port external traffic.

For ExternalName service, Istio generate:
 1) 0.0.0.0:port listener, direct to :
 2) STRICT_DNS type cluster, with 
 3) ExternalName specified external domain as endpoints 
 
Similar to ServiceEntry with MESH_EXTERNAL typed location and DNS typed resolution

**Comment by hzxuzhonghu** [comment]
+1  for this proposal: allocate an ip to externalName service, which could make externalName service behave like SE.

**Comment by howardjohn** [comment]
If there is desire to do this I would request it is behind the DNS_AUTO_ALLOCATE feature flag. This feature is inherently unsafe - tbh I regret that we added it...

**Comment by hzxuzhonghu** [comment]
Sure, make great sense

**Comment by craigbox** [comment]
https://twitter.com/ibuildthecloud/status/1684710979517685760

**Comment by keithmattix** [comment]
Fixed by #46332
